### DISCUSSION TITLE
[Feature] Miles & More / Star Alliance flight history import

### DISCUSSION THREAD

**Discussion by nicolaus-hee** [johanohly/AirTrail#155] [Feature Request]
Congrats on this app and thank you for making it!

### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I wanted to (1) share that an api endpoint containing your entire Star Alliance flight history in `json` format is available to people with a Lufthansa Miles & More account and (2) propose the feature of a corresponding AirTrail importer.

Steps to get the export:

1. Open `https://www.lufthansa.com/de/en/profile/travel-insights` and log in
2. Open `https://api.travelid.lufthansa.com//flightstats/v3/me/segmentList?departureDateRange=10000%20months&size=10000&page=0`

You'll get a response like so:

```
{
    "SegmentListResponses": [
        {
            "DepartureDate": "2024-11-22",
            "ArrivalDate": "2024-11-22",
            "OriginCityCode": "ZRH  ",
            "OriginCityName": "Zurich",
            "DestinationCityCode": "BER  ",
            "DestinationCityName": "Berlin",
            "OriginAirportCode": "ZRH",
            "OriginAirportName": "Zurich Airport",
            "DestinationAirportCode": "BER",
            "DestinationAirportName": "Berlin Brandenburg Apt",
            "StatusPoints": 20,
            "GupPoints": 20,
            "HonPoints": 0,
            "AirlineDesignatorCode": "LX   ",
            "FlightNumber": 962,
            "CompartmentClass": "M",
            "AircraftCode": "320  ",
            "Distance": 405,
            "TimeOnPlane": 5100,
            "StatusMiles": 0,
            "AwardMiles": 552,
            "DepartureTime": "2024-11-22T19:55:00.000+0000",
            "ArrivalTime": "2024-11-22T21:20:00.000+0000",
            "Honmiles": 0,
            "PnrrecordLocator": "VV5xxx  "
        },
        {
            "DepartureDate": "2024-11-18",
            "ArrivalDate": "2024-11-18",
            "OriginCityCode": "BER  ",
            "OriginCityName": "Berlin",
            "DestinationCityCode": "ZRH  ",
            "DestinationCityName": "Zurich",
            "OriginAirportCode": "BER",
            "OriginAirportName": "Berlin Brandenburg Apt",
            "DestinationAirportCode": "ZRH",
            "DestinationAirportName": "Zurich Airport",
            "StatusPoints": 20,
            "GupPoints": 20,
            "HonPoints": 0,
            "AirlineDesignatorCode": "LX   ",
            "FlightNumber": 981,
            "CompartmentClass": "M",
            "AircraftCode": "321  ",
            "Distance": 405,
            "TimeOnPlane": 5100,
            "StatusMiles": 0,
            "AwardMiles": 558,
            "DepartureTime": "2024-11-18T17:10:00.000+0000",
            "ArrivalTime": "2024-11-18T18:35:00.000+0000",
            "Honmiles": 0,
            "PnrrecordLocator": "VV5xxx  "
        }, ...........
]
```

In my case, that went back to 2015.

Thank you for your consideration.

**Comment by johanohly** [comment]
Thank you for making this feature request! 

I will definitely keep this on my radar for when I have more free time.

**Comment by EnricoVentola** [comment]
Working on it here @johanohly  @nicolaus-hee : https://github.com/EnricoVentola/AirTrail/tree/miles-and-more-import.

Using data I pulled from my Miles and More account, seat map class varies from website to website, but [Lufthansa say](https://developer.lufthansa.com/docs/read/api_partner/reference_data/Seat_Details)  **M** is Econony, **E** is Premium Economy (economy+), **C** is Business and **F** is First. 

My old paper tickets are a mish mash of codes like Y, L, O. This appears to be [Booking Class](https://www.flyertalk.com/forum/lufthansa-austrian-swiss-brussels-lot-other-partners-miles-more/657240-where-list-lh-booking-codes.html) not Compartment Class. Since we don't care about niche booking codes, I trusted the Lufthansa reference page, although maybe In the future we can include more specific breakdowns in the app?  [LH Booking Classes](https://expertworldtravel.com/lufthansa-booking-classes/)

This API reference got me optimistically thinking that since we can fetch the [seat design for LH aircraft](https://developer.lufthansa.com/docs/read/api_partner/reference_data) then instead of a text box for seat input for LH flights, we can show a visual display instead, like when you book flights online... one for later?

Some fields are blank if they are booked through Lufthansa but not Star Alliance members (Discover-4Y), occasionally quirky things missing for Turkish Airlines. (PnrrecordLocator, ArrivalTime, AircraftCode, DepartureTime). So, set the Zod to Optional and if the Times are missing let the app default to 00:00.

Couple of flights don't have ArrivalDate, so let's assume it arrived on the same date that it Departed, A check of [TK4008](https://www.flightaware.com/live/flight/THY4008/history/20240514/2105Z/LTAC/LTAI) and 7495, which were problematic for me shows that isn't always the case... but what else can we do??

If TimeOnPlane is present, then we can use that instead of running the differenceInSeconds calculation on Departure and ArrivalTime. 

Add Status, Hon and Award miles to the Notes if they have a value. [Last year LH moved to a new Points system](https://www.miles-and-more.com/de/en/program/status-benefits/new-statusprogramme/point-system.html), added that too if the value is > 0.

I think GupPoints and "Qualifying Points" and StatusPoints are just "Points". Added them both to the Notes section too.

We can find the Aircraft ICAO from AircraftCode, but, annoyingly, AircraftCode doesn't match ICAO very well:

- Airbus A320, A320, AircraftCode is 320
- Canada Regional CRJ-900, CRJ9, AircraftCode is CR9.
- Airbus A320neo, A20N, AircraftCode is 32N.
- Embraer E195 E2, E295, AircraftCode is E95.

I rushed out a method to convert these codes to ICAO in aircraft.ts, but it's messy and should be more re-usable.
